### PR TITLE
perf: optimize training loop

### PR DESCRIPTION
**Pull request by caic99**
Improvements to the training process:

* [`deepmd/pt/train/training.py`](diffhunk://#diff-a90c90dc0e6a17fbe2e930f91182805b83260484c9dc1cfac3331378ffa34935R659): Added a check to skip setting the model to training mode if it already is. The profiling result shows it takes some time to recursively set it to all models.

* [`deepmd/pt/train/training.py`](diffhunk://#diff-a90c90dc0e6a17fbe2e930f91182805b83260484c9dc1cfac3331378ffa34935L686-L690): Modified the gradient clipping function to include the `error_if_nonfinite` parameter, and removed the manual check for non-finite gradients and the associated exception raising.

## Summary by CodeRabbit

- **New Features**
	- Improved training loop with enhanced error handling and control flow.
	- Updated gradient clipping logic for better error detection.
	- Refined logging functionality for training and validation results.

- **Bug Fixes**
	- Prevented redundant training calls by adding conditional checks.

- **Documentation**
	- Clarified method logic in the `Trainer` class without changing method signatures.

### CODE REVIEW

File: deepmd/pt/train/training.py [hunk 1]
```diff
@@ -656,7 +656,8 @@ def step(_step_id, task_key="Default") -> None:
             # PyTorch Profiler
             if self.enable_profiler or self.profiling:
                 prof.step()
-            self.wrapper.train()
+            if not self.wrapper.training:
```
Comment (njzjz) [inline deepmd/pt/train/training.py:659]: I think it just checks the wrapper itself. Is there a situation that the wrapper is in the training mode but the model is in the eval mode?

Comment (njzjz) [reply deepmd/pt/train/training.py:659]: Btw, I don't think it should be put in the loop

Comment (caic99) [reply deepmd/pt/train/training.py:659]: > I think it just checks the wrapper itself. Is there a situation that the wrapper is in the training mode but the model is in the eval mode?

It seems that the status change only happens here
https://github.com/deepmodeling/deepmd-kit/blob/53bbbdbb890e14af79fc8e8a8de778ad3cc06750/deepmd/pt/train/training.py#L769
So I guess the answer should be no.

Comment (caic99) [reply deepmd/pt/train/training.py:659]: > Btw, I don't think it should be put in the loop

Agreed - maybe move it to the end of validation block?

Comment (njzjz) [reply deepmd/pt/train/training.py:659]: > move it to the end of validation block

Reasonable. Validation is not performed every step.

**Review by njzjz — APPROVED** [review]

**Comment by caic99** [comment]
Before:
image

After:

image
